Variables for artifacts

nojb opened the pull request:
This PR implements #2416, closes #2416, closes #1630, closes #1544

nojb commented:
So far the following are working:
- `%{EXT:MOD}` used inside a `dune` file where `EXT` is `cmo`, `cmx` or `cmi`, and `MOD` is a module name, and
- `%{EXT:LIB}` used inside a `dune` file where `EXT` is `cma` or `cmxa` and `LIB` is a module name

Next I will try to implement virtual targets to be used in the command line. This will make the whole thing very easy to test as well.

nojb commented:
This is ready for review.

In order to keep it as simple as possible (both for the user as for the implementation), I made a series of simplifications with respect to the discussion in #2416, which I find fit nicely with the purpose of these variables and the fact that they are intended for advanced users:

- The variables do not care if the object being referred to belongs to a different project or if it is private, you are free to refer to modules and/or libraries in arbitrary directories of the workspace.
- Each library is referred to by its local name, not the public one, and the variables resolve to paths in the build context (`_build/default/...`), not the installation one.
- The error handling in the case the module/library does not exist is a bit unsatisfactory (see the tests for examples): the reason is that current expansion API does not allow to propagate an error that occurs during expansion of a macro. At first glance it looks like improving on this would require some non-trivial refactoring.

One other interesting change is that I now use the parser for the dependency specification to parse the command line (in the non-alias case), opening the door to `dune build '(glob_files *.foo)'`, etc. But all this is left for a later PR.

nojb commented on test/blackbox-tests/test-cases/variables-for-artifacts/run.t:
```diff
@@ -0,0 +1,123 @@
+  26 | (alias
+  27 |  (name t3)
+  28 |  (deps %{cmo:xxxx}))
+  Error: No rule found for xxxx.cmo
```
When the object being referred to (here, a module named `xxxx` in the current directory), the macro expands anyway to a fictitious filename (here, `xxxx.cmo`) which then fails but with a message that is a bit misleading and does not mention the initial macro.

The problem is that it is not currently possible (as far as I can see) to propagate an error that occurs during expansion of a macro.

diml replied:
Do you mean to propagate the location?

diml replied:
nervermind, I see what you mean. Currently, the best is to error out while expanding the macro

diml replied:
Actually, yh, we do need to allow reporting proper error when expanding macros. The result of `expand_var` is actually a bit misleading. We return `Error <expansion>` for things that need to be expanded later. Instead of returning a `result`, we should return something like this:

```
type expanded = Static of Value.t list | Dynamic of Pform.Expansion.t | Error of User_message.t
```

nojb replied:
I am trying to see how far the error case needs to be propagated inside the existing infrastructure. 

Does it sound reasonable that the result of `String_with_vars.expand` would be a `(Value.t list, User_message.t) result` or maybe even `(Value.t, User_message.t) result list` ?

diml replied:
That seems right. We could also use the `Or_exn.t` type here since we use it for the same purpose in other places. Not too sure which of the two types you propose is better though.

nojb replied:
Thanks for the hint, will look into it.

nojb replied:
I think this is now fixed, thanks.

nojb commented on test/blackbox-tests/test-cases/variables-for-artifacts/run.t:
```diff
@@ -0,0 +1,123 @@
+  [1]
+
+  $ dune build %{cmo:xxxx}
+  Error: Don't know how to build xxxx.cmo
```
Here you see another instance of the issue.

nojb commented on test/blackbox-tests/test-cases/variables-for-artifacts/run.t:
```diff
@@ -0,0 +1,123 @@
+
+  $ dune build %{cma:bar} %{cmxa:a1}
+  Error: Don't know how to build bar.cma
+  Hint: did you mean a1.cma?
```
And here, where the hint is unfortunate.

nojb commented on test/blackbox-tests/test-cases/variables-for-artifacts/run.t:
```diff
@@ -0,0 +1,190 @@
+
+  $ cd sub && dune build --display short %{cmx:x}
+  Info: Creating file dune-project with this contents:
+  | (lang dune 1.11)
```
I don't understand why invoking `dune` from the `sub` subdirectory makes it create a `dune-project` file there; there is already one in the parent directory.

diml replied:
When running dune inside dune, we disable root detection. It's like passing `--root .`. It's to avoid the inner dune from crawling up the filesystem outside of the test.

nojb replied:
Got it, thanks.

nojb commented on src/expander.ml:
```diff
@@ -97,12 +112,54 @@ let expand_var_exn t var syn =
+      | Some lookup_module -> begin
+          match lookup_module ~dir (Module.Name.of_string name) with
+          | None ->
+            Some [Value.String (s ^ Pform.Artifact.ext a)]
```
Here we return a fictitious filename when the module being referred to does not exist, which will cause an error later (albeit with a misleading error message). Same for libraries below.

nojb commented on bin/target.ml:
```diff
@@ -106,9 +106,41 @@ let resolve_path path ~(setup : Dune.Main.build_system) =
+              (Dune.Dir_contents.artifacts (Dune.Dir_contents.get sctx ~dir)) name
+          in
+          let expander = Dune.Expander.set_lookup_module expander ~lookup_module in
+          let expander = Dune.Expander.set_lookup_library expander ~lookup_library in
```
It would be better to set these lookup functions inside the function `Super_context.expander` but `Dir_contents` depends on `Super_context` so there is a dependency issue.

diml commented:
Could you add some doc in the manual for this feature?

diml commented on bin/target.ml:
```diff
@@ -107,9 +107,41 @@ let resolve_path path ~(setup : Dune.Main.build_system) =
+            Dune.Dune_file.Dep_conf.decode
+        in
+        Dune_lang.Decoder.parse parser Univ_map.empty
+          (Dune_lang.parse_string ~fname:"command line" ~mode:Dune_lang.Parser.Mode.Single s)
```
Could you move this parsing into a cmdliner `Arg.conv` value? i.e. like the `path` value in `bin/arg.ml`. The idea would be to have:

```ocaml
val Arg.dep : Dune.Dune_file.Dep_conf.t Arg.conv
```

Then the error would be reported by cmdliner while parsing the command line, which seems nicer to me.

Since the parsing reports `User_error.E` exception, you will need to extract the error message from the `User_message.t` value. You can use this (from `.../sexp_tests.ml`) for that:

```ocaml
let extract_error_message (msg : User_message.t) =
  Format.asprintf "%a"
    Pp.render_ignore_tags
    (User_message.pp { msg with loc = None })
  |> String.drop_prefix ~prefix:"Error: "
  |> Option.value_exn
  |> String.trim
```

nojb replied:
Good idea, will do.

nojb replied:
This is now done.

nojb commented:
> Could you add some doc in the manual for this feature?

Done. I think all other issues have been addressed so this is again ready for review :)

diml commented on bin/arg.ml:
```diff
@@ -25,3 +25,98 @@ let path = Path.conv
+          match parse_alias s with
+          | None -> x
+          | Some (true, s) -> alias_rec s
+          | Some (false, s) -> alias s
```
I believe this function is parsing `(file @foo)` as `Alias_rec "foo"`. We should probably call `parse_alias` before

nojb replied:
You are right, I had missed that; will fix.

nojb replied:
This is fixed now, thanks.

diml commented:
Ok, I started the review but I won't have time to finish it today. I'll continue on Monday except if someone else does it before that :)

rgrinberg commented on bin/arg.ml:
```diff
@@ -25,3 +25,92 @@ let path = Path.conv
+      with
+      | x -> `Ok x
+      | exception (User_error.E msg) ->
+        `Error (User_message.to_string msg)
```
Why can't we rely on Report_error to print this error?

nojb commented:
I have split off the cmdliner part of this into a different PR for easier reviewing #2500.

rgrinberg commented:
@nojb once this is rebased I will review ASAP

nojb commented:
> @nojb once this is rebased I will review ASAP

@rgrinberg rebased and ready for review. I cleaned up the history as well, so now even commit-by-commit review is possible :) Thanks!

nojb commented:
Rebased and ready for review :)

rgrinberg commented on src/dune/expander.ml:
```diff
@@ -90,40 +107,102 @@ let expand_env t pform s : Value.t list option =
+  let dir = Path.Build.parent_exn path in
+  match a with
+  | Pform.Artifact.Mod kind -> (
+    match t.lookup_module with
```
Can we use `Option.O` and `let*` here?

nojb replied:
Yes, done.

rgrinberg commented on src/dune/expander.ml:
```diff
@@ -90,40 +107,102 @@ let expand_env t pform s : Value.t list option =
+          | Some path ->
+              Some (Ok [ Value.Path (Path.build path) ]) ) ) )
+  | Lib mode -> (
+    match t.lookup_library with
```
Ditto

nojb replied:
Done.

rgrinberg commented on src/dune/expander.ml:
```diff
@@ -90,40 +107,102 @@ let expand_env t pform s : Value.t list option =
+         | Macro (Artifact a, s) -> (
+             let loc = String_with_vars.Var.loc var in
+             match expand_artifact ~dir ~loc t a s with
+             | Some (Ok v) ->
```
Let's use `Option.map` or `let+`.

nojb replied:
Done.

rgrinberg commented on src/dune/expander.ml:
```diff
@@ -1,5 +1,12 @@
+type expanded =
+  | Static of Value.t list
+  | Dynamic of Pform.Expansion.t
+  | Error of User_message.t
```
What about:

```
type expanded = Static of .. | Dynamic of ..
```

and then use:

```
(expanded, User_message.t) Result.t
```

After working a lot of Lwt, I'm not a huge of fan inlining the error case in any type.

nojb replied:
OK, done.

nojb commented:
> Okay, this is good enough to merge. One thing that I'm still questioning is the usefulness of this `Obj_dir.Module.Single.t` type. First of all, is it really that much better than a tuple? But really, is it even necessary? We can always go from module name to stanza (exe/library) and from that we can construct the object directory. This isn't essential, but I'm usually not a fan of storing stuff that we can easily calculate it. Especially if it's inexpensive.

I agree; I removed `Obj_dir.Module.Single` and switched to a tuple instead.

nojb commented:
Planning to merge this after CI passes.

nojb commented:
Fixes #2416 #1630 #1544

nojb commented:
Actually, there is still an outstanding issue : currently these variables cannot be used in the `action` field of `rule`. The issue is that `rule` is interpreted early as a side effect of computing the `Dir_contents.t` in order to extract their targets. But I need to have the `Dir_contents.t` already computed in order to expand the variables.

I don't think there is any fundamental problem here, as the variables do not declare targets so it is "just" a matter of delaying their expansion until after the `Dir_contents.t` has been computed. But it looks like this will require some refactoring.

I am looking int this. If it looks like it requires some work, maybe we can merge this as-is, since it already adds some new functionality (command-line convenience targets), and then do the remaining work on a follow-up PR.

nojb commented:
@diml @rgrinberg I would appreciate your opinion on the last commit (`[RFC] user rules...`).

The idea is to separate the computation of the targets of `rule` from the installation of the corresponding rule in the context. This allows passing different expanders at each point and allows to use the variables in this PR in the `action` field of `rule` (as explained in the previous note, the issue here is that in order to expand these variables we need access to the full `Dir_contents.t`, but the targets of the `rule` are needed *while* building this value).

Right now the I am duplicating the setup of the build arrow for `rule`, which is not satisfactory. Do you see how to memoize to avoid the duplicated work? Or perhaps a different approach altogether?

diml commented:
Would it be enough to make the expansion of the artifact variables dynamic? That should break the cycle

nojb commented:
> Would it be enough to make the expansion of the artifact variables dynamic? That should break the cycle

Thanks for the hint! I gave it a try (commit "Expand artifact variables dynamically") and now it works, but am not 100% sure about the implementation. Could you take a look and let me know if it looks like it is in the right direction?

rgrinberg commented on test/blackbox-tests/test-cases/variables-for-artifacts/dune:
```diff
@@ -0,0 +1,84 @@
+;; (rule
+;;  (targets my2.cmxs)
+;;  (deps %{cmx:x2})
+;;  (action (run %{ocamlopt} -shared -o %{targets} %{cmxa:plugin} %{cmx:x3})))
```
You might as well make this a separate failing test I believe. Should be quite useful down the line and this way we won't forget to uncomment it if someone adds support

nojb replied:
Good idea, will do.

rgrinberg commented:
I approved because I don't think it's a good idea to delay this PR further. We can resolve whatever issues come up in separate PR's. Let's just make sure the test suite is extensive.

nojb commented:
> I approved because I don't think it's a good idea to delay this PR further. We can resolve whatever issues come up in separate PR's. Let's just make sure the test suite is extensive.

OK, planning to merge once the CI passes. Thanks!

nojb commented:
Merged. Thanks everyone for the help!

diml commented:
> Thanks for the hint! I gave it a try (commit "Expand artifact variables dynamically") and now it works, but am not 100% sure about the implementation. Could you take a look and let me know if it looks like it is in the right direction?

I just had a quick look and it looks good to me!
